### ISSUE TITLE
Allow environment variables to be set on build basis + ondemand builds

### ISSUE THREAD

**Issue by themihai** [drone/drone#894]
I would like to propose a feature and also to offer my help to implement/develop it if it is accepted. The feature is important for organisations with many repositories and many teams. It addresses the security of environment variables.
Feature:
  Allow the client to run builds on arbitrary branches/commits with specific environment variables so that it doesn't need to expose secrets in config files or web ui.

Example:
``curl -X POST https://drone.io/api/repos/github.com/foo/bar/run?branch=develop&env[0]=[secret_name=secret_value]``
  A CLI could be used to automate this kind of request(e.g. run it from the current branch and it picks up the branch and repository) so you may need just one more command after push. 
The command above would run a new build using the latest changes of the develop branch with a secret set as environment variable(secret_name=secret_value). I find this transparent and more secure than defining the secrets in config files or the web UI. 
Use case:
   Many projects require external dependencies during the build process. An example is a go project which imports many other repositories/packages. The external dependencies have specific access restrictions (e.g. only some developers have access to them). The develop has no problem to download these dependencies locally because it's supposed to already have access to them (though github teams ACL or whatever). However on the build server various secrets/tokens are required(tokens, password etc). 
The current practice is to simply define the secrets/tokens as environment variables. Usually the token is global because the tokens management becomes laborious as the number of dependencies and repositories increases. This presents a great security risk because a rough team member(or someone who hacked the account)  with access to a small repository within the organisation could simply commit code to print the secrets/tokens in the build log and use them to access restricted dependencies. This may even go unnoticed because it may commit the code in an unmonitored branch so the change/build is not reviewed.
  Another issue is that even if you would set a different secret/token for each repository(unfeasible)  there is still the risk for someone who leaves the team to use the secrets later when its access was restricted. It's not feasible to change all the tokens whenever someone leaves a specific team.

**Comment by bradrydzewski** [comment]
@themihai  I definitely appreciate you are bringing solutions to the table. Do you think we could take a step back, however, and focus more on defining specific problems that we would like to solve? I think having a clear problem statement(s) is beneficial before we get to the solution :)

**Comment by themihai** [comment]
@bradrydzewski here is the specific problem: How can I prevent a rough developer to access secret/tokens required by the build/tests (e.g. Amazon S3 credentials to download/upload various assets) ?

**Comment by bradrydzewski** [comment]
@themgt thanks for clarifying. We are working on a few changes that I think will start to address this:

First, we are going to use a separate container to clone all repository data into a volume. We just started working on this, early code here: https://github.com/drone/drone-clone-git

Second, we are going to use separate containers to execute all plugins. Plugin containers will link directly to the same volume mentioned above. So you could imagine an S3 container that has access to the volume in order to upload artifacts to S3.

Both the above changes are important because we no longer need to pass the environment variables directly to the build container. This would (for example) prevent a malicious team member from dumping secret data to the logs.

Third, for some plugins we allow you to set global configuration data in the `drone.toml` configuration file. This isn't available for all plugins, but we are doing this with the `npm` plugin:
https://github.com/drone/drone/blob/master/plugin/publish/npm/npm.go#L43

There are still edge cases that need to be considered, but I think we are definitely taking steps in the right direction.

**Comment by themihai** [comment]
It definitely seems to address the issue and sounds exciting but it seems to me that the solution works only with a  limited set of plugins (e.g. the S3 plugin, npm) but many build systems need custom scripting to resolve the dependencies or to deploy.

**Comment by bradrydzewski** [comment]
I think this could be resolved in a few different ways. The first, and most important, is to use pull requests and perform code reviews before merging code.

Off the top of my head, here are some ideas to add security to deployments:

1. Create a custom plugin that fails when a commit is introduced that changes the deployment script
2. Create a tightly controlled custom plugin to do your deployments instead of using a deployment script that can be altered by a committer
3. Setup a VPN or list of restricted IP addresses for accessing servers. Only Drone should be able to access these servers, so that if the developer gets the credentials they cannot access from their machine

And some ideas to add more security to cloning dependencies:

1. Use [machine users](https://developer.github.com/guides/managing-deploy-keys/#machine-users) when setting up your builds, in conjunction with the proposed feature in #61. When an employee leaves you would only need to revoke, re-generate and register a single key.
2. Use GitHub Enterprise behind the firewall. If an employee is terminated and restricted from the network they will be unable to access GitHub rendering stolen keys useless

These are just some ideas off the top of my head and I'm definitely open to suggestions ...

I would also like to point out that if a developer can make malicious changes to your build or deployment script, and **nobody** notices, what is to prevent them from making malicious changes to your application code? And then that application eventually being executed in production?

I think the above point really highlights the importance of strong code reviews

**Comment by vaijab** [comment]
This is not an easy thing to get right. I'll start by explaining my use case.

I am using drone to run a CI against a platform build. It is bunch of scripts, unit files, aws stacks and so on. There is no way I could use drone plugins to do deployments.

So my CI run looks like this:
- run some lint tests
- run some validation tests
- generate some files / certs / etc
- create an s3 bucket
- create cloudformation stacks
- run some more tests

All of the above bits run in the `build` step in `.drone.yml`, I cannot use plugins because there are too many custom scripts. Currently the only bit of secrets I need is `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`, but there might be more.

@bradrydzewski how would you go about providing secrets for this?

**Comment by bradrydzewski** [comment]
@vaijab you can inject into the build section like this:

```
build:
  commands:
    - go build
    - go test
    - ./deploy --secret=$$AWS_SECRET_ACCESS_KEY
```

or if you don't want the secret to get printed in your output you can do this:

```
build:
  commands:
    - go build
    - go test
  environment:
    - AWS_SECRET_ACCESS_KEY =$$AWS_SECRET_ACCESS_KEY
```

Note that secrets won't get injected into the `build` section for pull requests at the moment, for security reasons. That is a separate issue that can be tracked at https://github.com/drone/drone/issues/1250

**Comment by vaijab** [comment]
@bradrydzewski thanks for that. That's exactly what I am doing right now, but I am seeing that secrets aren't getting injected for PRs. I just wanted to check whether I am on the same page.

Here is a situation:

I branch off master which contains `.drone.sec` with some secrets in it. I make some changes to dump secrets into a build log, drone runs, I grab all the secrets. What I don't understand how disabling secret injection at PR hook helps?

**Comment by bradrydzewski** [comment]
> What I don't understand how disabling secret injection at PR hook helps?

Let's use the Drone project as an example. If I inject `AWS_SECRET_ACCESS_KEY ` into the build section of my `.drone.yml` a developer could send me a pull request with this code:

```
println(os.Getenv($AWS_SECRET_ACCESS_KEY"))
```

The above code would execute and write my credentials to the build output, making it available for the whole world to see and steal.

While this may not apply to all repositories, it definitely applies to public repositories. The current stable version of Drone takes a very conservative approach because this functionality is new and untested in the real world. We do plan to loosen the policy in subsequent releases as we learn more about the different use cases and how people use Drone. Note issue #1250, referenced above.

**Comment by vaijab** [comment]
> While this may not apply to all repositories, it definitely applies to public repositories.

Oh I see. Makes sense now. Thanks

**Comment by bradrydzewski** [comment]
I wanted to provide some notes in case someone is interested in sending a pull request, as this should be relatively straight forward to add.

We have an existing endpoint to re-start / fork a build: 
https://github.com/drone/drone/blob/master/server/build.go#L155

We can alter this endpoint to accept a set of variables via url query parameters:

```diff
func PostBuild(c *gin.Context) {

	remote_ := remote.FromContext(c)
	repo := session.Repo(c)
	fork := c.DefaultQuery("fork", "false")

+	var params = map[string]string{}
+	for key, val := range c.Request.URL.Query() {
+	if key == "fork" || key == "event" || key == "deploy_to" { continue }
+		params[key]=val
+	}
```

Then we can set these query parameters in the build job environment section later on in that same function:

```diff
	for _, job := range jobs {
		job.Error = ""
		job.Status = model.StatusPending
		job.Started = 0
		job.Finished = 0
		job.ExitCode = 0
		job.NodeID = 0
		job.Enqueued = build.Enqueued
		store.UpdateJob(c, job)

+		for key, val := range params {
+			job.Environment[key]=val
+		}
	}
```

and then finally it probably makes sense to update the `drone build start` command to include the ability to pass variables via the command line.

**Comment by gtaylor** [comment]
I don't have the capacity to implement this right now, but I might be able to assist with testing this in the next week or two if an implementation arises. Feel free to @-mention me.

**Comment by vaijab** [comment]
I'll give this a shot. Hope to have something working in the next couple of days.

**Comment by bradrydzewski** [comment]
@vaijab awesome thanks. Hit me up in gitter if you have any questions

**Comment by bradrydzewski** [comment]
merged #1735 which gives the ability to attach parameters to an on-demand build:
https://github.com/drone/drone/issues/894#issuecomment-233707244

this will be available once 0.5 is released

**Comment by cyrillay** [comment]
Hi @bradrydzewski ,
Currently it is possible to pass arguments (target deployment environment) when using the drone promoting feature, however, it doesn't seem to be possible to pass environment variables, please correct me if I'm wrong.

The use cases i'm trying to solve are the following : 
- One application can contain multiple 'jobs' that are deployed the exact same way, but it is currently impossible to deploy them separately, without having one pipeline step for each job
- One application can be deployed the same way, but we want to pass some configuration at runtime (the url to expose on, or a URL to connect to for example)

I am currently doing this by having multiple deployment steps (triggered by promotion events) for each different deployment and the different configurations/environment variables are hardcoded for each different step.

Please let me know if that makes sense, or if there is another way of solving this.

Thanks !
